File: cucumber-expressions/javascript/test/combinatorial_generated_expression_factory_test.js
/* eslint-env mocha */
const assert = require('assert')
const ParameterType = require('../src/parameter_type')
const CombinatorialGeneratedExpressionFactory = require('../src/combinatorial_generated_expression_factory')

describe('CucumberExpressionGenerator', () => {
  it('generates multiple expressions', () => {
    const parameterTypeCombinations = [
      [
        new ParameterType(
          'color',
          /red|blue|yellow/,
          null,
          s => s,
          false,
          true
        ),
        new ParameterType(
          'csscolor',
          /red|blue|yellow/,
          null,
          s => s,
          false,
          true
        ),
      ],
      [
        new ParameterType(
          'date',
          /\d{4}-\d{2}-\d{2}/,
          null,
          s => s,
          false,
          true
        ),
        new ParameterType(
          'datetime',
          /\d{4}-\d{2}-\d{2}/,
          null,
          s => s,
          false,
          true
        ),
        new ParameterType(
          'timestamp',
          /\d{4}-\d{2}-\d{2}/,
          null,
          s => s,
          false,
          true
        ),
      ],
    ]

    const factory = new CombinatorialGeneratedExpressionFactory(
      'I bought a {%s} ball on {%s}',
      parameterTypeCombinations
    )
    const expressions = factory.generateExpressions().map(ge => ge.source)
    assert.deepEqual(expressions, [
      'I bought a {color} ball on {date}',
      'I bought a {color} ball on {datetime}',
      'I bought a {color} ball on {timestamp}',
      'I bought a {csscolor} ball on {date}',
      'I bought a {csscolor} ball on {datetime}',
      'I bought a {csscolor} ball on {timestamp}',
    ])
  })
})

File: cucumber-expressions/ruby/spec/cucumber/cucumber_expressions/combinatorial_generated_expression_factory_test.rb
# frozen_string_literal: true

require 'cucumber/cucumber_expressions/parameter_type'
require 'cucumber/cucumber_expressions/combinatorial_generated_expression_factory'

module Cucumber
  module CucumberExpressions
    class Color; end
    class CssColor; end
    class Date; end
    class DateTime; end
    class Timestamp; end

    describe CombinatorialGeneratedExpressionFactory do
      it 'generates multiple expressions' do
        parameter_type_combinations = [
          [
            ParameterType.new('color', /red|blue|yellow/, Color, ->(_s) { Color.new }, true, false),
            ParameterType.new('csscolor', /red|blue|yellow/, CssColor, ->(_s) { CssColor.new }, true, false)
          ],
          [
            ParameterType.new('date', /\d{4}-\d{2}-\d{2}/, Date, ->(_s) { Date.new }, true, false),
            ParameterType.new('datetime', /\d{4}-\d{2}-\d{2}/, DateTime, ->(_s) { DateTime.new }, true, false),
            ParameterType.new('timestamp', /\d{4}-\d{2}-\d{2}/, Timestamp, ->(_s) { Timestamp.new }, true, false)
          ]
        ]

        factory = CombinatorialGeneratedExpressionFactory.new(
          'I bought a {%s} ball on {%s}',
          parameter_type_combinations
        )
        expressions = factory.generate_expressions.map(&:source)
        expect(expressions).to eq([
                                    'I bought a {color} ball on {date}',
                                    'I bought a {color} ball on {datetime}',
                                    'I bought a {color} ball on {timestamp}',
                                    'I bought a {csscolor} ball on {date}',
                                    'I bought a {csscolor} ball on {datetime}',
                                    'I bought a {csscolor} ball on {timestamp}'
                                  ])
      end
    end
  end
end
